---
title: "MAF files summary"
author: "UMCCR"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    theme: readable
    toc: true
    toc_float: true
    code_folding: hide
  rmdformats::material:
    highlight: kate
params:
  maf_dir: 
  maf_files: 
  datasets: 
  genes_min: 
  genes_list: 
  genes_blacklist: 
  samples_list:
  samples_blacklist: 
  nonSyn_list: 
  out_folder: 'MAF_summary_report'
---

Mutation data summary for dataset(s) **`r gsub(",", ", ", params$datasets) `**

***

**NOTE**: this report summarises **only non-synonymous variants** defined as variants with the following consequences: *`r params$nonSyn_list`*.

<details>
<summary>Variants consequence impact definitions</summary>
<font size="2">

* [*High impact variant consequence*](http://asia.ensembl.org/Help/Glossary?id=535){target="_blank"} -	the variant is assumed to have high (disruptive) impact in the protein, probably causing protein truncation, loss of function or triggering nonsense mediated decay.

* [*Moderate impact variant consequence*](http://asia.ensembl.org/Help/Glossary?id=535){target="_blank"} -	a non-disruptive variant that might change protein effectiveness.

* [*Low impact variant consequence*](http://asia.ensembl.org/Help/Glossary?id=535){target="_blank"} -	a variant that is assumed to be mostly harmless or unlikely to change protein behaviour.

Variants with high/moderate variant consequences include: *frame shift deletions*, *frame shift deletions*, *splice site mutations*, *translation start site mutations*, *nonsense mutation*, *nonstop mutations*, *in-frame deletion*, *in-frame insertions* and *missense mutation*.

</font> 
</details>

***

```{r define_functions, comment=NA, message=FALSE, warning=FALSE}
##### Define functions

##### Create 'not in' operator
"%!in%" <- function(x,table) match(x,table, nomatch = 0) == 0

##### Assign colours to different datasets. These colours will be used to distinguish tabs in generated excel summary spreadsheets
getDatasetsColours <- function(datasets) {
  
  ##### Predefined selection of colours for datasets
  datasets.colours <- c("dodgerblue","firebrick","lightslategrey","darkseagreen","orange","darkcyan","bisque", "coral2", "cadetblue3","red","blue","green")
  
  f.datasets <- factor(datasets)
  vec.datasets <- datasets.colours[1:length(levels(f.datasets))]
  datasets.colour <- rep(0,length(f.datasets))
  for(i in 1:length(f.datasets))
    datasets.colour[i] <- vec.datasets[ f.datasets[i]==levels(f.datasets)]
  
  return( list(vec.datasets, datasets.colour) )
}

###### Generate dataTable for each dataset with all mutation information for selected gene(s), as provided in MAF files
mut.details.datasets <- function(mafInfo, datasets, genes) {
  
  ##### Vector with datasets with no mutations reported in selected genes
  datasets.noMut <- NULL
  
  ##### Create a list for htmlwidgets
  widges.list <- htmltools::tagList()
  
  for ( i in 1:length(datasets) ) {
  
    mut.details.genes <- mafInfo[[datasets[i]]]@data[ mafInfo[[datasets[i]]]@data[, Hugo_Symbol] %in% genes, ]
    
    if ( nrow(mut.details.genes) != 0 ) {
      
      ##### Sort table by gene symbol and then by sample ID
      mut.details.genes <- mut.details.genes[ order(mut.details.genes$Hugo_Symbol, mut.details.genes$Tumor_Sample_Barcode), ]
      
      widges.list[[i]] <- DT::datatable( data = mut.details.genes, caption = htmltools::tags$caption(style = 'caption-side: top; text-align: left;', htmltools::strong(datasets[i])), filter = "top", extensions = c('Buttons','FixedColumns','Scroller'), options = list(pageLength = 10, dom = 'Bfrtip', buttons = c('excel', 'csv', 'pdf','copy','colvis'), scrollX = TRUE, fixedColumns = list(leftColumns = 2), deferRender = TRUE, scrollY = 200, scroller = TRUE) ) %>%
        DT::formatStyle( columns = names(mut.details.genes), 'text-align' = 'center' )
      
    } else {
      datasets.noMut <- c(datasets.noMut, datasets[i])
    }
  }
  
  ##### Report datasets with no mutations reported in selected genes
  if ( length(datasets.noMut) != 0 ) {
    
    cat(paste("Note, queried gene(s)", genes, "have no non-synonymous variants reported in the following dataset(s):", paste(datasets.noMut, collapse = ", "), "\n\n\n\n", sep=" "))
  }
  
  ##### Print a list of htmlwidgets
  widges.list
}

###### Generate lollipop plot for each dataset for selected gene
lollipops.datasets <- function(mafInfo, datasets, gene) {
  
  ##### Create a list to store MAF info for individual datasets
  for ( dataset in datasets ) {
    
    ##### Check if the gene has any mutations in correspoding dataset
    if ( nrow(subsetMaf(maf = mafInfo[[dataset]], includeSyn = TRUE, genes = gene, 
        query = "Variant_Type != 'CNV'")) != 0 ) {
      
      ##### Drawing lollipop for the top 10 genes in each dataset
      ##### Check if the amino acid changes information is available in MAF provided files. The script expects column called "HGVSp_Short", which is produced with vcf2maf (https://github.com/mskcc/vcf2maf) when converting VCFs to MAFs (https://github.com/cBioPortal/cbioportal/issues/2996) and describes a mutation's amino acid change. The "aa_mutation" field used for annotation in ICGC samples is also acceptable. NOTE: other possibilities are: "Protein_Change", "AAChange""
       if ( "HGVSp_Short" %in% maftools::getFields(mafInfo[[dataset]]) && gene %in% mafInfo[[dataset]]@data$Hugo_Symbol ) {
        
        cat(paste("\n\n <b>", dataset, "</b> \n\n", sep=" "))
        
        ##### Make it plot to a dummy graphics device file (e.g. /dev/null) to avoid plotting to the console
        pdf(file="/dev/null")
        lollipopPlot.image <- capture.output(maftools::lollipopPlot(maf = mafInfo[[dataset]], gene = gene, AACol = "HGVSp_Short", printCount = FALSE, showDomainLabel = FALSE, repel = FALSE, labelPos = "all" , showMutationRate = TRUE, cBioPortal = TRUE, fn = paste0(mutationMapsDir, "/", paste(dataset, gene, sep="_"))))
        invisible(dev.off())
        
        ##### Export pdf to png
        lollipopPlot.image <- image_read_pdf(paste(mutationMapsDir, "/", paste(dataset, gene, sep="_"), ".pdf", sep = ""), pages = NULL, density = 300)
        image_write(lollipopPlot.image, path = paste(mutationMapsDir, "/", paste(dataset, gene, sep="_"), ".png", sep = ""), format = "png")
        
        ##### Read in the oncoplots PNG files
        cat("![](",paste(paste0(mutationMapsDir, "/", paste(dataset, gene, sep="_")), ".png", sep = ""),")")
        
        ##### Remove redundant pdf plot
        file.remove(paste(mutationMapsDir, "/", paste(dataset, gene, sep="_"), ".pdf", sep = ""))
    
      } else if ( "aa_mutation" %in% maftools::getFields(mafInfo[[dataset]]) && gene %in% mafInfo[[dataset]]@data$Hugo_Symbo ) {
        
        cat(paste("\n\n <b>", dataset, "</b> \n\n", sep=" "))
        
        ##### Make it plot to a dummy graphics device file (e.g. /dev/null) to avoid plotting to the console
        pdf(file="/dev/null")
        lollipopPlot.image <- capture.output(maftools::lollipopPlot(maf = mafInfo[[dataset]], gene = gene, AACol = "aa_mutation", printCount = FALSE, showDomainLabel = FALSE, repel = FALSE, labelPos = "all" , showMutationRate = TRUE, cBioPortal = TRUE, fn = paste0(mutationMapsDir, "/", paste(dataset, gene, sep="_"))))
        invisible(dev.off())
        
        ##### Export pdf to png
        lollipopPlot.image <- image_read_pdf(paste(mutationMapsDir, "/", paste(dataset, gene, sep="_"), ".pdf", sep = ""), pages = NULL, density = 300)
        image_write(lollipopPlot.image, path = paste(mutationMapsDir, "/", paste(dataset, gene, sep="_"), ".png", sep = ""), format = "png")
        
        ##### Read in the oncoplots PNG files
        cat("![](",paste(paste0(mutationMapsDir, "/", paste(dataset, gene, sep="_")), ".png", sep = ""),")")
        
        ##### Remove redundant pdf plot
        file.remove(paste(mutationMapsDir, "/", paste(dataset, gene, sep="_"), ".pdf", sep = ""))
      
      ##### ...otherwise leave a message
      } else {
        
        ##### Check if the genes has any synonymous vatiants
        if ( "HGVSp_Short" %in% names(mafInfo[[dataset]]@maf.silent) && gene %in% mafInfo[[dataset]]@maf.silent$Hugo_Symbol  ) {
        
          cat(paste("This section was skipped for dataset", dataset, "since only synonymous variants were detected in", gene, "gene.\n\n\n\n", sep=" "))
      
        } else if ( "aa_mutation" %in% names(mafInfo[[dataset]]@maf.silent) && gene %in% mafInfo[[dataset]]@maf.silent$Hugo_Symbol  ) {
          
          cat(paste("This section was skipped for dataset", dataset, "since only synonymous variants were detected in", gene, "gene.\n\n\n\n", sep=" "))
      
        } else {
          cat(paste("This section was skipped for dataset", dataset, "since the corresponding MAF does not contain field with amino acid changes details!\n\n\n\n", sep=" "))
        }
      }
        
      } else {
        
      cat(paste("\n\n <b>", dataset, "</b> \n\n", sep=" "))
      cat(paste("Gene <i>", gene, "</i> has no mutations reported in dataset", dataset, "\n\n\n\n", sep=" "))
    }
  }
}

##### Stop the maftools::tcgaCompare function from printing the summary table
tcgaCompare_mod <- function (maf, capture_size = NULL, cohortName = NULL, primarySite = FALSE, 
    col = c("gray70", "black"), medianCol = "red", fn = NULL, 
    width = 8, height = 5, fontSize = 10) 
{
    tcga.cohort = system.file("extdata", "tcga_cohort.txt.gz", 
        package = "maftools")
    if (Sys.info()[["sysname"]] == "Windows") {
        tcga.cohort.gz = gzfile(description = tcga.cohort, open = "r")
        tcga.cohort <- suppressWarnings(data.table(read.csv(file = tcga.cohort.gz, 
            header = TRUE, sep = "\t", stringsAsFactors = FALSE)))
        close(tcga.cohort.gz)
    }
    else {
        tcga.cohort = data.table::fread(input = paste("zcat <", 
            tcga.cohort), sep = "\t", stringsAsFactors = FALSE)
    }
    if (primarySite) {
        tcga.cohort = tcga.cohort[, .(Tumor_Sample_Barcode, total, 
            site)]
        colnames(tcga.cohort)[3] = "cohort"
    }
    else {
        tcga.cohort = tcga.cohort[, .(Tumor_Sample_Barcode, total, 
            cohort)]
    }
    maf.mutload = getSampleSummary(maf)[, .(Tumor_Sample_Barcode, 
        total)]
    if (is.null(cohortName)) {
        cohortName = "Input"
    }
    maf.mutload[, `:=`(cohort, cohortName)]
    tcga.cohort$total = as.numeric(as.character(tcga.cohort$total))
    maf.mutload$total = as.numeric(as.character(maf.mutload$total))
    if (!is.null(capture_size)) {
        maf.mutload[, `:=`(total, total/capture_size)]
        tcga.cohort[, `:=`(total, total/50)]
    }
    tcga.cohort = rbind(tcga.cohort, maf.mutload)
    tcga.cohort.med = tcga.cohort[, .(.N, median(total)), cohort][order(V2, 
        decreasing = TRUE)]
    tcga.cohort$cohort = factor(x = tcga.cohort$cohort, levels = tcga.cohort.med$cohort)
    colnames(tcga.cohort.med) = c("Cohort", "Cohort_Size", "Median_Mutations")
    tcga.cohort$TCGA = ifelse(test = tcga.cohort$cohort %in% 
        cohortName, yes = "Input", no = "TCGA")
    tcga.cohort.gg = ggplot(data = tcga.cohort, aes(x = cohort, 
        y = total, group = Tumor_Sample_Barcode, color = TCGA)) + 
        geom_point(position = position_jitter(width = 0.3), size = 0.2, 
            alpha = 0.8) + geom_point(data = tcga.cohort.med, 
        aes(y = Median_Mutations, x = Cohort), shape = 95, inherit.aes = FALSE, 
        color = medianCol, size = 6) + cowplot::theme_cowplot(font_size = fontSize, 
        line_size = 1) + cowplot::background_grid(major = "x") + 
        theme(axis.text.x = element_text(angle = 45, vjust = 1, 
            hjust = 1, face = "bold"), legend.position = "none", 
            axis.text.y = element_text(face = "bold", size = 10), 
            axis.title.y = element_text(face = "bold", size = 12)) + 
        scale_color_manual(values = rev(col))
    if (is.null(capture_size)) {
        tcga.cohort.gg = tcga.cohort.gg + scale_y_log10(breaks = c(0.01, 
            0.1, 1, 10, 100, 1000, 10000)) + ylab("log10 (Mutations per sample)")
    }
    else {
        tcga.cohort.gg = tcga.cohort.gg + expand_limits(y = c(0.01, 
            1000)) + scale_y_log10(breaks = c(0.01, 0.1, 1, 10, 
            100, 1000), labels = c(0.01, 0.1, 1, 10, 100, 1000)) + 
            ylab("log10 (Mutations per MB)")
    }
    if (!is.null(fn)) {
        cowplot::save_plot(filename = paste0(fn, ".pdf"), plot = tcga.cohort.gg, 
            base_height = height, base_width = width)
        write.table(tcga.cohort.med, file = paste0(fn, "_mutation_load.tsv"), 
            sep = "\t", quote = FALSE, row.names = FALSE)
    }
    #message("Summary..")
    #print(tcga.cohort.med)
    print(tcga.cohort.gg)
    return(tcga.cohort.gg)
}
```

```{r load_libraries, warning=FALSE}
suppressMessages(library(knitr))
suppressMessages(library(maftools))
suppressMessages(library(openxlsx))
suppressMessages(library(ggplot2))
suppressMessages(library(DT))
suppressMessages(library(magick))
```

## Datasets

This report summarises and visualises mutation data in [Mutation Annotation Format](https://software.broadinstitute.org/software/igv/MutationAnnotationFormat){target="_blank"} (MAF) files for the following dataset(s):

```{r datasets, comment = NA}
##### Present patient cohorts to be summarised
##### Split the string of MAF files and put them into a vector
mafFiles <- unlist(strsplit(params$maf_files, split=',', fixed=TRUE))
mafFiles <- paste(params$maf_dir, mafFiles, sep="/")

##### Split the string of datasets names and put them into a vector
datasets.list <- unlist(strsplit(params$datasets, split=',', fixed=TRUE))

datasets.df <- as.data.frame( cbind(datasets.list, unlist(strsplit(params$maf_files, split=',', fixed=TRUE))) )
names(datasets.df) <- c("Dataset", "MAF file")

DT::datatable( data = datasets.df, filter = "none", extensions = 'Buttons', options = list(pageLength = length(mafFiles), dom = 'Bfrtip', buttons = c('excel', 'csv', 'pdf','copy')) ) %>%
        DT::formatStyle( columns = names(datasets.df), 'text-align' = 'center' )
```

```{r load_data, comment = NA, message=FALSE, warning=FALSE, results='hide'}
##### Read MAF files and put associated info into a list
##### Create a list to store MAF info for individual datasets
mafInfo <- vector("list", length(mafFiles))
names(mafInfo) <- datasets.list

##### NOTE: maftools by default summarises only non-synonymous variants with high/moderate variant consequences and ignores silent variants (https://github.com/PoisonAlien/maftools/issues/63), which are stored in "maf.silent" slot of the class MAF object (mafInfo[[i]]@maf.silent)
for ( i in 1:length(mafFiles) ) {
  mafInfo[[i]] <- maftools::read.maf(maf = mafFiles[i], vc_nonSyn = params$nonSyn_list, verbose = FALSE)
}

##### Create directory for output files
outDir <- paste(params$maf_dir, params$out_folder, sep = "/")
if ( !file.exists(params$out_folder) ){

  dir.create(outDir, recursive=TRUE)
}

##### Read in list of genes of interest of specified
if ( !is.na(params$genes_list) ){
  
  goi <- unique(read.table(params$genes_list, sep="\t", as.is=TRUE, header=FALSE, row.names=NULL)[,1])
}
```

```{r include_exclude_samples, comment = NA, message=FALSE, warning=FALSE}
##### Include user-derined samples(s) for the analysis
if ( !is.na(params$samples_list) ) {
    
  for ( i in 1:length(mafFiles) ) {
      
    ##### Read in list of samples to be excluded
    inclsamples.df <- read.table(params$samples_list, sep="\t", as.is=TRUE, header=TRUE, row.names=NULL)
    samples2keep <- unique(inclsamples.df[,"Tumor_Sample_Barcode"])
    
    ##### Subset the maf object to include only use-defined sample(s)
    ##### Initially don't save the subset output as maf object, as this will exlude samples with no non-synonymous mutations from the summary
    mafInfo[[i]] <- subsetMaf(maf = mafInfo[[i]], tsb = as.vector(samples2keep), genes = NULL, fields = NULL, query = NULL, mafObj = FALSE, includeSyn = TRUE, dropLevels=TRUE)
    
    ##### Now read the data subset as a maf object
    mafInfo[[i]] <- maftools::read.maf(maf = mafInfo[[i]], vc_nonSyn = params$nonSyn_list, verbose = FALSE)
  }
}

if ( !is.na(params$samples_blacklist) ) {
    
  for ( i in 1:length(mafFiles) ) {
      
    ##### Read in list of samples to be excluded
    exclsamples.df <- read.table(params$samples_blacklist, sep="\t", as.is=TRUE, header=TRUE, row.names=NULL)
    exclsamples <- unique(exclsamples.df[,"Tumor_Sample_Barcode"])
    
    samples2keep <- unlist(unique( mafInfo[[i]]@maf.silent[, "Tumor_Sample_Barcode"]))[ unlist(unique(mafInfo[[i]]@maf.silent[, "Tumor_Sample_Barcode"])) %!in% exclsamples ]
    
    ##### Subset the maf object to exclude use-defined sample(s)
    ##### Initially don't save the subset output as maf object, as this will exlude samples with no non-synonymous mutations from the summary
    mafInfo[[i]] <- subsetMaf(maf = mafInfo[[i]], tsb = as.vector(samples2keep), genes = NULL, fields = NULL, query = NULL, mafObj = FALSE, includeSyn = TRUE, dropLevels=TRUE)
    
    ##### Now read the data subset as a maf object
    mafInfo[[i]] <- maftools::read.maf(maf = mafInfo[[i]], vc_nonSyn = params$nonSyn_list, verbose = FALSE)
  }
}
```

```{r exclude_genes, comment = NA, message=FALSE, warning=FALSE}
##### Exclude user-derined gene(s) from the analysis
if ( !is.na(params$genes_blacklist) ) {
    
  for ( i in 1:length(mafFiles) ) {
    
    ##### Read in list of genes to be excluded
    exclgenes <- unique(read.table(params$genes_blacklist, sep="\t", as.is=TRUE, header=FALSE, row.names=NULL)[,1])
    
    genes2keep <- unique( mafInfo[[i]]@data$Hugo_Symbol)[ unique( mafInfo[[i]]@data$Hugo_Symbol) %!in% exclgenes ]
  
    ##### Subset the maf object to exclude use-defined gene(s)
    ##### Initially don't save the subset output as maf object, as this will exlude samples with no non-synonymous mutations from the summary
    mafInfo[[i]] <- subsetMaf(maf = mafInfo[[i]], tsb = NULL, genes = genes2keep, fields = NULL, query = NULL, mafObj = FALSE, includeSyn = TRUE)
    
    ##### Now read the data subset as a maf object
    mafInfo[[i]] <- maftools::read.maf(maf = mafInfo[[i]], vc_nonSyn = params$nonSyn_list, verbose = FALSE)
  }
}
```

```{r silent_variants, comment = NA, message=FALSE, warning=FALSE}
##### Identify and record samples with no non-synonymous mutations
##### Prepare list to store all samples and samples with > 0 non-synonymous variants
MAF_samples <- vector("list", length(datasets.list))
names(MAF_samples) <- datasets.list
MAF_samples.silent.df <- NULL

##### Loop through MAF files
for ( i in 1:length(mafFiles) ) {
  
  ##### Identify samples with no non-synonymours variants according to corresponding MAF file
  MAF_samples[[i]]$all <- unlist(unique(mafInfo[[i]]@maf.silent[, "Tumor_Sample_Barcode"]))
  MAF_samples[[i]]$nonsyn <- unlist(maftools::getSampleSummary(mafInfo[[i]])[, "Tumor_Sample_Barcode"])
  MAF_samples[[i]]$silent <-  MAF_samples[[i]]$all[ MAF_samples[[i]]$all %!in% MAF_samples[[i]]$nonsyn  ]
  
  ##### Check if there thre are any samples with no non-synonymours variants. If so, add them to data frame
  if ( length(MAF_samples[[i]]$silent) > 0 ) {
    for ( sample in MAF_samples[[i]]$silent ) {
      
      MAF_samples.silent.df <- rbind( MAF_samples.silent.df, cbind( datasets.list[i], sample))
    }
    colnames(MAF_samples.silent.df) <- c("Dataset", "Sample")
  }
}
```

***

## Summary tables {.tabset}

### Overall summary

Table(s) with basic information about each dataset based on data in corresponding MAF file(s).

```{r overll_summary, comment = NA, message=FALSE, warning=FALSE}
##### Write overall summary into a file
##### Assign different colour to individual datasets
datasets.colour <- getDatasetsColours(datasets.list)

##### Create a new workbook
wb <- createWorkbook("MAF_summary.xlsx")

##### Add worksheets, one for each dataset
for ( i in 1:length(mafFiles) ) {
  
    addWorksheet(wb, substring(datasets.list[i], 0, 31), tabColour = datasets.colour[[1]][i])
    writeData(wb, sheet = i, mafInfo[[i]]@summary)
}

saveWorkbook(wb, paste(outDir, "MAF_summary.xlsx", sep="/"), overwrite = TRUE)

##### Present a MAF file summary table in the html report
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

for ( i in 1:length(mafFiles) ) {
  widges.list[[i]] <- DT::datatable( data = mafInfo[[i]]@summary, caption = htmltools::tags$caption(style = 'caption-side: top; text-align: left;', htmltools::strong(datasets.list[i])), filter = "none", extensions = 'Buttons', options = list(pageLength = nrow(mafInfo[[i]]@summary), dom = 'Bfrtip', buttons = c('excel', 'csv', 'pdf','copy','colvis')) ) %>%
        DT::formatStyle( columns = names(mafInfo[[i]]@summary), 'text-align' = 'center' )
}

##### Print a list of htmlwidgets
widges.list

##### Add extra lines to make sure that this section doesn't overlap with the next one
cat("\n\n\n")
```

```{r maf_fields, warning=FALSE}
##### Additionally ceate an excel spreadsheet listing all fields (columns) in the individaul MAF files
##### Create a new workbook
wb <- createWorkbook("MAF_fields.xlsx")

##### Add worksheets, one for each dataset
for ( i in 1:length(mafFiles) ) {
  
    addWorksheet(wb, substring(datasets.list[i], 0, 31), tabColour = datasets.colour[[1]][i])
    writeData(wb, sheet = i, maftools::getFields(mafInfo[[i]]))
}

saveWorkbook(wb, paste(outDir, "MAF_fields.xlsx", sep="/"), overwrite = TRUE)
```

***

### Samples summary {.tabset}

#### Samples with non-synonymous variant(s)

Table(s) summarising samples in individual datasets. Each table contains per-sample information (rows) about *number of different types of mutations* (columns), as well as the *total number of mutations* reported in corresponding MAF file. Note, only **samples with detected non-synonymous variant(s)** are reported in the table below.

```{r sample_summary, comment = NA, warning=FALSE}
##### Write samples summary into a file
##### Create a new workbook
wb <- createWorkbook("MAF_sample_summary.xlsx")

##### Add worksheets, one for each dataset
for ( i in 1:length(mafFiles) ) {
  
    addWorksheet(wb, substring(datasets.list[i], 0, 31), tabColour = datasets.colour[[1]][i])
    writeData(wb, sheet = i, maftools::getSampleSummary(mafInfo[[i]]))
}

saveWorkbook(wb, paste(outDir, "MAF_sample_summary.xlsx", sep="/"), overwrite = TRUE)

##### Present a sample table in the html report
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

for ( i in 1:length(mafFiles) ) {
  widges.list[[i]] <- DT::datatable( data = maftools::getSampleSummary(mafInfo[[i]]), caption = htmltools::tags$caption(style = 'caption-side: top; text-align: left;', htmltools::strong(datasets.list[i])), filter = "top", extensions = c('Buttons','Scroller'), options = list(pageLength = 10, dom = 'Bfrtip', buttons = c('excel', 'csv', 'pdf','copy','colvis'), scrollX = TRUE, deferRender = TRUE, scrollY = 200, scroller = TRUE) ) %>%
        DT::formatStyle( columns = names(maftools::getSampleSummary(mafInfo[[i]])), 'text-align' = 'center' )
}

##### Print a list of htmlwidgets
widges.list

##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()
```

***

#### Samples with no non-synonymous variants

The table below lists sample(s) in which **no non-synonymous variants** were detected and hence will not be included in the summary tables/plots.

```{r sample_no_nonsynonymous, comment = NA, warning=FALSE}
##### report samples with no non-synonymous variants according to corresponding MAF file
if ( !is.null(MAF_samples.silent.df) ) {

  DT::datatable( data = MAF_samples.silent.df, caption = htmltools::tags$caption(style = 'caption-side: top; text-align: left;', htmltools::strong("Samples with no non-synonymours variants detected")), filter = "top", extensions = 'Buttons', options = list(pageLength = 10, dom = 'Bfrtip', buttons = c('excel', 'csv', 'pdf','copy')) ) %>%
        DT::formatStyle( columns = names(MAF_samples.silent.df), 'text-align' = 'center' )
  
} else {
  cat("Non-synonymous variants were detected in all samples.\n\n\n")
}

##### Add extra lines to make sure that this section doesn't overlap with the next one
cat("\n\n\n")
```

***

#### `r if ( !is.na(params$samples_blacklist) ) { c("Excluded samples") }`

List of samples excluded from the analysis.

```{r excluded_samples_table, comment = NA, message=FALSE, warning=FALSE}
##### Present a samples table in the html report
if ( !is.na(params$samples_blacklist) ) {
    
  DT::datatable(data = exclsamples.df, caption = htmltools::tags$caption(style = 'caption-side: top; text-align: left;'), filter = "top", extensions = c('Buttons','FixedColumns','Scroller'), options = list(pageLength = 10, dom = 'Bfrtip', buttons = c('excel', 'csv', 'pdf','copy','colvis'), scrollX = TRUE, fixedColumns = list(leftColumns = 1), deferRender = TRUE, scrollY = 200, scroller = TRUE) ) %>%
      DT::formatStyle( columns = names(exclsamples.df), 'text-align' = 'center' )
  
} else {
  cat("No samples were excluded.\n\n\n")
}
```

***

### Genes summary {.tabset}

#### Mutated genes

Table(s) summarising mutated genes in individual datasets. Each table contains per-gene information (rows) about *number of different types of mutations* (columns), as well as the *total number of mutations* reported in corresponding MAF file. The last two columns contain the *number of samples with mutations/alterations* in the corresponding gene.

```{r gene_summary, warning=FALSE}
##### Write gene summary into a file
##### Create a new workbook
wb <- createWorkbook("MAF_gene_summary.xlsx")

##### Add worksheets, one for each dataset
for ( i in 1:length(mafFiles) ) {
  
    addWorksheet(wb, substring(datasets.list[i], 0, 31), tabColour = datasets.colour[[1]][i])
    writeData(wb, sheet = i, maftools::getGeneSummary(mafInfo[[i]]))
}

saveWorkbook(wb, paste(outDir, "MAF_gene_summary.xlsx", sep="/"), overwrite = TRUE)

##### Present a gene table in the html report
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

for ( i in 1:length(mafFiles) ) {
  widges.list[[i]] <- DT::datatable(data = maftools::getGeneSummary(mafInfo[[i]]), caption = htmltools::tags$caption(style = 'caption-side: top; text-align: left;', htmltools::strong(datasets.list[i])), filter = "top", extensions = c('Buttons','FixedColumns','Scroller'), options = list(pageLength = 10, dom = 'Bfrtip', buttons = c('excel', 'csv', 'pdf','copy','colvis'), scrollX = TRUE, fixedColumns = list(leftColumns = 2), deferRender = TRUE, scrollY = 200, scroller = TRUE) ) %>%
        DT::formatStyle( columns = names(maftools::getGeneSummary(mafInfo[[i]])), 'text-align' = 'center' )
}

##### Print a list of htmlwidgets
widges.list
```

***

#### Excluded genes

List of genes excluded from the analysis.

```{r excluded_genes_table, comment = NA, message=FALSE, warning=FALSE}
##### Present a gene table in the html report
if ( !is.na(params$genes_blacklist) ) {
    
  DT::datatable(data = data.frame(Gene = exclgenes), caption = htmltools::tags$caption(style = 'caption-side: top; text-align: left;'), filter = "top", extensions = c('Buttons','FixedColumns','Scroller'), options = list(pageLength = 10, dom = 'Bfrtip', buttons = c('excel', 'csv', 'pdf','copy'), scrollX = TRUE, deferRender = TRUE, scrollY = 200, scroller = TRUE) ) %>%
    DT::formatStyle( columns = "Gene", 'text-align' = 'center' )
  
} else {
  cat("No genes were excluded.\n\n\n")
}
```


```{r top_genes_threshold, warning=FALSE}
##### Check if the use-defined number of genes to present is not higher than the number of genes with non-synonymous mutations reported in individual MAFs
top_genes_no <- vector("list", length(mafFiles))
names(top_genes_no) <- datasets.list

mut_freq <- vector("list", length(mafFiles))
names(mut_freq) <- datasets.list

##### Set height for oncoplots
oncoplot_height <-  vector("list", length(mafFiles))
names(oncoplot_height) <- datasets.list

for ( i in 1:length(mafFiles) ) {
  
  ##### Get mutation frequency for each gene
  mut_freq[[i]] <- round(getGeneSummary(mafInfo[[i]])$AlteredSamples/as.numeric(mafInfo[[i]]@summary[3,"summary"])*100, digits=0)
  names(mut_freq[[i]]) <- getGeneSummary(mafInfo[[i]])$Hugo_Symbol
  
  ##### Get the number of genes that have mutations in the defined minimal percentage of samples (4% as default)
  top_genes_no[[i]] <- length( mut_freq[[i]][ mut_freq[[i]] >= as.numeric(params$genes_min) ] )
  
  ##### Set height for oncoplots
  if ( top_genes_no[[i]] > 100 ) {
    
    oncoplot_height[[i]] <- 100*top_genes_no[[i]]/3
    
  } else if ( top_genes_no[[i]] > 19 ) {
    
    oncoplot_height[[i]] <- 100*top_genes_no[[i]]/2
    
  } else {
    oncoplot_height[[i]] <- 900
  }
}
```

```{r goi, warning=FALSE}
##### Deal with additional genes of interest (if specified)
##### Check if the use-defined genes of interest are already present withing the top most frequently mutated genes
if ( !is.na(params$genes_list) ) {
  
  genes_list_goi <- vector("list", length(mafFiles))
  names(genes_list_goi) <- datasets.list
  
  goi_absent <- vector("list", length(mafFiles))
  names(goi_absent) <- datasets.list
  
  top_genes_goi <- vector("list", length(mafFiles))
  names(top_genes_goi) <- datasets.list
  
  top_genes <- vector("list", length(mafFiles))
  names(top_genes) <- datasets.list
  
  ##### Set height for oncoplots
  oncoplot_goi_height <-  vector("list", length(mafFiles))
  names(oncoplot_goi_height) <- datasets.list
  
  for ( i in 1:length(mafFiles) ) {
    
    ##### Extract the user-defined number of top mutated genes
    top_genes[[i]] <- maftools::getGeneSummary(mafInfo[[i]])$Hugo_Symbol[1:top_genes_no[[i]]]
    
    ##### Record genes of interest that are not present in the MAF file
    goi_absent[[i]] <- goi[ goi %!in% maftools::getGeneSummary(mafInfo[[i]])$Hugo_Symbol ]
    
    ##### Keep the genes of interest and most frequently mutated genes separately
    top_genes_goi[[i]] <- goi[ goi %in% top_genes[[i]] ]
    genes_list_goi[[i]] <- goi[ goi %!in% top_genes[[i]] ]
    
    ##### Remove genes absent in MAF from the genes of interest list
    genes_list_goi[[i]] <- genes_list_goi[[i]][ genes_list_goi[[i]] %!in% goi_absent[[i]] ]

    ##### Set height for oncoplots
    if ( length(genes_list_goi[[i]]) > 100 ) {
      
      oncoplot_goi_height[[i]] <- 100*length(genes_list_goi[[i]])/3
      
    } else if  ( length(genes_list_goi[[i]]) > 19 ) {
      
      oncoplot_goi_height[[i]] <- 100*length(genes_list_goi[[i]])/2
      
    } else {
      oncoplot_goi_height[[i]] <- 900
    }
  }
}

##### Add extra lines to make sure that this section doesn't overlap with the next one
cat("\n\n\n")
```

***

## Summary plots {.tabset .tabset-fade}

### Oncoplot {.tabset .tabset-fade}

#### Recurrently mutated genes

Oncoplot(s) illustrating different types of mutations observed across samples for the **most frequently mutated genes** (mutated in at least `r params$genes_min` % of patients). The side- and top bar-plots present the frequency of mutations in these genes and in individual samples, respectively.

```{r maf_oncoplot, comment = NA, results="asis"}
##### Save oncoplot into the file
##### Generate separate plot for each dataset
for ( i in 1:length(mafFiles) ) {
  
  cat(paste("\n\n <b>", datasets.list[i], "</b> \n\n", sep=" "))
  cat(paste(top_genes_no[[i]], "genes are mutated in at least", params$genes_min, "% of patients\n\n", sep=" "))
  
  ##### Generate oncoplots for the top mutated genes in each dataset that has > 1 sample with non-synonymous variants
  if ( top_genes_no[[i]] > 1 ) {
    
    ##### Save the plot as PNG
    png( file = paste(outDir, "/MAF_oncoplot_", datasets.list[i], ".png", sep = ""), width = 1800, height = oncoplot_height[[i]], units = "px", res = 200 )
      
    ##### Drawing oncoplots for the top mutated genes in each dataset
    plot.new()
    par(mar=c(4,4,2,0.5), oma=c(1.5,2,2,1))
    maftools::oncoplot(maf = mafInfo[[i]], top = top_genes_no[[i]], fontSize = 12, colbar_pathway = TRUE, removeNonMutated = FALSE)
    dev.off()
    
    ##### Read in the oncoplots PNG files
    cat("![](",paste(outDir, "/MAF_oncoplot_", datasets.list[i], ".png", sep = ""),")")
    cat(paste(rep("-", 80)), "\n")
    
  } else {
      cat(paste0("No genes are mutated in at least", params$genes_min, "% of patients.\n\n\n"))
  }
}
```

***

#### Additional genes of interest

Oncoplot(s) illustrating different types of mutations observed across samples for the **genes of interest** (if specified). The side and top bar-plots present the frequency of mutations in these genes and in individual samples, respectively.

```{r maf_oncoplot_goi, comment = NA, results="asis"}
##### Draw additional oncoplot if a list of genes of intereset was specified by user 
###### Generate separate plot for each dataset
if ( !is.na(params$genes_list) && length(goi) > 1 ) {
  
  for ( i in 1:length(mafFiles) ) {
  
    cat(paste("\n\n <b>", datasets.list[i], "</b>\n\n", sep=" "))
    
    ##### Drawing oncoplots for the top mutated genes in each dataset that has > 1 sample with non-synonymous variants
    if ( length(genes_list_goi[[i]]) > 1 ) {
    
      ##### Report genes missing in individual MAFs, as well as those which are within the list of genes of interest and are also among the most frequently mutated genes
      cat(paste("* The following genes from provided gene list are missing from", datasets.list[i], "MAF:\n\n", sep=" "))
      
      cat(paste("<i>", goi_absent[[i]], "</i>", collapse = ", "))
      
      cat(paste("\n\n* The following genes from provided gene list are also among the top", top_genes_no[[i]], "most frequently mutated genes and are ignored in this oncoplot:\n\n", sep=" "))
      
      cat(paste("<i>", top_genes_goi[[i]], "</i>", collapse = ", "), "\n\n")
      
      ##### Save the plot as PNG
      png( file = paste(outDir, "/MAF_oncoplot_goi_", datasets.list[i], ".png", sep = ""), width = 1800, height = oncoplot_goi_height[[i]], units = "px", res = 200 )
        
      ##### Drawing oncoplots for the top mutated genes in each dataset
      plot.new()
      par(mar=c(4,4,2,0.5), oma=c(1.5,2,2,1))
      maftools::oncoplot(maf = mafInfo[[i]], genes = genes_list_goi[[i]], fontSize = 12, colbar_pathway = TRUE, removeNonMutated = FALSE)
      dev.off()
      
      ##### Read in the oncoplots PNG files
      cat("![](",paste(outDir, "/MAF_oncoplot_goi_", datasets.list[i], ".png", sep = ""),")")
      cat(paste(rep("-", 80)), "\n")

    } else {
      cat(paste0("All genes of interest are among the top ", top_genes_no[[i]], " most frequently mutated genes (mutated in at least", params$genes_min, "% of patients) or no non-synonymous variants were detected in ", datasets.list[i], " dataset.\n\n\n"))
    }
  }
} else {
  cat("No list with genes of interest was provided\n\n\n\n")
}
```

***

### MAF summary

A per-MAF file summary including *frequency of various mutation/SNV types/classes* (top panel), the *number of variants in each sample* as a stacked bar-plot (bottom-left) and *variant types* as a box-plot (bottom-middle), as well as the *frequency of different mutation types* for the **`r top_genes_no[[i]]` most frequently mutated genes** (mutated in at least `r params$genes_min` % of patients) (bottom-right). The horizontal dashed line in stacked bar-plot represents median number of variants across the dataset.

```{r maf_summary_plot, comment = NA, fig.width = 9, fig.height = 10}
###### Generate separate plot for each dataset
for ( i in 1:length(mafFiles) ) {

  cat(paste(datasets.list[i], "\n\n", sep=" "))
  
  ##### Plotting MAF summary
  par(mar=c(4,4,2,0.5), oma=c(1.5,2,2,1))
  maftools::plotmafSummary(maf = mafInfo[[i]], top = top_genes_no[[i]], rmOutlier = TRUE, addStat = 'median', dashboard = TRUE, titvRaw = FALSE)
  mtext("MAF summary", outer=TRUE,  cex=1, line=-0.5)
}
```

***

### Transitions/transversions {.tabset .tabset-fade}

#### Plots {.tabset .tabset-fade}

Plots presenting the transitions and transversions distribution in individual dataset(s). In each panel, the box-plots show the *overall distribution* of the six different conversions (*C>A*, *C>G*, *C>T*, *T>C*, *T>A* and *T>G*)(top-left), and the transitions and transversions *frequency* (top-right). The stacked bar-plot (bottom) displays the *fraction* of the six different conversions in each sample.

```{r maf_TiTv_plot, comment = NA}
###### Generate separate plot for each dataset
###### Create empty list for TiTv info from each dataset
titv.info <- list()

for ( i in 1:length(mafFiles) ) {

  cat(paste(datasets.list[i], "\n\n", sep=" "))

  ##### Drawing distribution plots of the transitions and transversions
  titv.info[[datasets.list[i]]] <- maftools::titv(maf = mafInfo[[i]], plot = FALSE, useSyn = TRUE)

  try(maftools::plotTiTv(res = titv.info[[datasets.list[i]]]), silent = TRUE)
  mtext("Transition and transversions distribution", outer=TRUE,  cex=1, line=-1.5)
}
```

***

#### Tables {.tabset .tabset-fade}

##### Overall distribution

Table(s) presenting the *overall distribution* of the six different conversions (*C>A*, *C>G*, *C>T*, *T>C*, *T>A* and *T>G*) across all samples in each dataset.

```{r maf_TiTv_tables_xlsx}
##### Write per-sample transitions and transversions distribution into a file
##### Create a new workbook
wb <- createWorkbook("MAF_summary_titv.xlsx")

##### Add worksheets, three for each dataset
for ( i in 1:length(mafFiles) ) {
  
    addWorksheet(wb, substring(paste0(datasets.list[i], " (fraction)"), 0, 31), tabColour = datasets.colour[[1]][i])
    writeData(wb, sheet = i*3-2, titv.info[[datasets.list[i]]]$fraction.contribution)
    
    addWorksheet(wb, substring(paste0(datasets.list[i], " (count)"), 0, 31), tabColour = datasets.colour[[1]][i])
    writeData(wb, sheet = i*3-1, titv.info[[datasets.list[i]]]$raw.counts)
    
    addWorksheet(wb, substring(paste0(datasets.list[i], " (TiTv fractions)"), 0, 31), tabColour = datasets.colour[[1]][i])
    writeData(wb, sheet = i*3, titv.info[[datasets.list[i]]]$TiTv.fractions)
}

saveWorkbook(wb, paste(outDir, "MAF_summary_titv.xlsx", sep="/"), overwrite = TRUE)
```

```{r maf_TiTv_table_overall_dist}
##### Generate tables with the overall distribution of the six different conversions (C>A, C>G, C>T, T>C, T>A and T>G)
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

for ( i in 1:length(mafFiles) ) {
  widges.list[[i]] <- DT::datatable(data = titv.info[[datasets.list[i]]]$raw.counts, caption = htmltools::tags$caption(style = 'caption-side: top; text-align: left;', htmltools::strong(datasets.list[i])), filter = "top", extensions = c('Buttons','Scroller'), options = list(pageLength = 10, dom = 'Bfrtip', buttons = c('excel', 'csv', 'pdf','copy','colvis'), scrollX = TRUE, deferRender = TRUE, scrollY = 200, scroller = TRUE) ) %>%
        DT::formatStyle( columns = names(titv.info[[datasets.list[i]]]$raw.counts), 'text-align' = 'center' )
}

##### Print a list of htmlwidgets
widges.list

##### Add extra lines to make sure that this section doesn't overlap with the next one
cat("\n\n\n")
```

***

##### Transitions/transversions frequency

Table(s) presenting the transitions and transversions *frequency* across all samples in each dataset.

```{r maf_TiTv_table_frequency}
##### Generate tables with transitions and transversions frequency in individual datasets
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

for ( i in 1:length(mafFiles) ) {
  widges.list[[i]] <- DT::datatable(data = titv.info[[datasets.list[i]]]$TiTv.fractions, caption = htmltools::tags$caption(style = 'caption-side: top; text-align: left;', htmltools::strong(datasets.list[i])), filter = "top", extensions = c('Buttons','Scroller'), options = list(pageLength = 10, dom = 'Bfrtip', buttons = c('excel', 'csv', 'pdf','copy','colvis'), scrollX = TRUE, deferRender = TRUE, scrollY = 200, scroller = TRUE) ) %>%
        DT::formatStyle( columns = names(titv.info[[datasets.list[i]]]$TiTv.fractions), 'text-align' = 'center' ) %>%
  formatRound(columns = c("Ti", "Tv"), 1)
}

##### Print a list of htmlwidgets
widges.list

##### Add extra lines to make sure that this section doesn't overlap with the next one
cat("\n\n\n")
```

***

##### Transitions/transversions fraction

Table(s) presenting the *fraction* of the six different conversions in each sample in individual dataset(s).

```{r maf_TiTv_table_fraction}
##### Generate tables with transitions and transversions frequency in individual datasets
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

for ( i in 1:length(mafFiles) ) {
  widges.list[[i]] <- DT::datatable(data = titv.info[[datasets.list[i]]]$fraction.contribution, caption = htmltools::tags$caption(style = 'caption-side: top; text-align: left;', htmltools::strong(datasets.list[i])), filter = "top", extensions = c('Buttons','Scroller'), options = list(pageLength = 10, dom = 'Bfrtip', buttons = c('excel', 'csv', 'pdf','copy','colvis'), scrollX = TRUE, deferRender = TRUE, scrollY = 200, scroller = TRUE) ) %>%
        DT::formatStyle( columns = names(titv.info[[datasets.list[i]]]$fraction.contribution), 'text-align' = 'center' ) %>%
  formatRound(columns = names(titv.info[[datasets.list[i]]]$fraction.contribution)[2:7], 1)
}

##### Print a list of htmlwidgets
widges.list

##### Add extra lines to make sure that this section doesn't overlap with the next one
cat("\n\n\n")
```

***

### Cross-cancers comparison

Plot(s) illustrating the mutation load in investigated dataset(s) along with distribution of variants compiled from over 10,000 whole-exome sequencing samples across 33 [TCGA](https://cancergenome.nih.gov/){target="_blank"} landmark cohorts. Every dot represents a sample whereas the red horizontal lines are the median numbers of mutations in the respective cancer types. The vertical axis (log scaled) shows the number of mutations per megabase whereas the different cancer types are ordered on the horizontal axis based on their median numbers of somatic mutations. This plot is similar to the one described in the paper [Signatures of mutational processes in human cancer](https://www.ncbi.nlm.nih.gov/pubmed/23945592){target="_blank"} by Alexandrov *et al*.

```{r maf_tcga_cohorts, comment = NA, message=FALSE, warning=FALSE}
###### Generate separate plot for each dataset
for ( i in 1:length(mafFiles) ) {
  
  cat(paste(datasets.list[i], "\n\n", sep=" "))
  
  ##### Compare mutation load against TCGA cohorts
  tcgaCompare_mod(maf = mafInfo[[i]], cohortName = datasets.list[i], primarySite=TRUE)
}
```

***

## Mutation maps {.tabset .tabset-fade}

Lollipop plot(s) showing mutation spots on protein structure for the **30 most frequently mutated genes** across all datasets. Many oncogenes have a preferential sites which are mutated more often than any other locus. These spots are considered to be mutational hot-spots and lollipop plots can be used to display them along with rest of the mutations. Note, that for the MAF file does not contain field with amino acid changes details. The longest transcript is used if multiple transcripts are available.

```{r top_mutated_genes, comment = NA, message=FALSE, warning=FALSE}
##### Get the 30 most frequently mutated genes across all datasets
mutRate.top.datasets <- NULL

for ( i in 1:length(mafFiles) ) {
  
  genes.top <- maftools::getGeneSummary(mafInfo[[i]])[c(1:top_genes_no[[i]]), Hugo_Symbol]
  
  ##### Add genes of interest to the list (if specified)
  if ( !is.na(params$genes_list) ){
    
    genes.top <- c(genes.top, genes_list_goi[[i]])
  }
    
  sampleSize <- as.numeric(mafInfo[[i]]@summary[ID %in% "Samples", summary])
  
  mutRate.top <- round(getGeneSummary(x = mafInfo[[i]])[ Hugo_Symbol %in% genes.top, MutatedSamples]/sampleSize * 100, digits = 2)
  names(mutRate.top) <- genes.top
  
  mutRate.top.datasets <- c( mutRate.top.datasets, mutRate.top)
}

mutRate.top.datasets <- sort(sort(mutRate.top.datasets, decreasing = TRUE)[unique(names(mutRate.top.datasets))], decreasing = TRUE)[c(1:top_genes_no[[i]])]

##### Create directory for pdf files
mutationMapsDir <- paste0(normalizePath(outDir), "/", "MAF_mutation_maps")

if ( !file.exists(mutationMapsDir) ){
  dir.create(mutationMapsDir, recursive=TRUE)
}
```

```{r prot_structre, comment = NA, message=FALSE, warning=FALSE}
##### Get list of proteins for which structure is available within maftools
gff = system.file('extdata', 'protein_domains.RDs', package = 'maftools')
gff = readRDS(file = gff)
```

### `r if ( !is.na(names(mutRate.top.datasets)[1]) ) { names(mutRate.top.datasets)[1] }`

```{r lollipop_plot_1, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 1st top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[1]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[1]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[1], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[2]) ) { names(mutRate.top.datasets)[2] }`

```{r lollipop_plot_2, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 2nd top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[2]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[2]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[2], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[3]) ) { names(mutRate.top.datasets)[3] }`

```{r lollipop_plot_3, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 3rd top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[3]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[3]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[3], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[4]) ) { names(mutRate.top.datasets)[4] }`

```{r lollipop_plot_4, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 4th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[4]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[4]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[4], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[5]) ) { names(mutRate.top.datasets)[5] }`

```{r lollipop_plot_5, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 5th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[5]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[5]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[5], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[6]) ) { names(mutRate.top.datasets)[6] }`

```{r lollipop_plot_6, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 6th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[6]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[6]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[6], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[7]) ) { names(mutRate.top.datasets)[7] }`

```{r lollipop_plot_7, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 7th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[7]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[7]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[7], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[8]) ) { names(mutRate.top.datasets)[8] }`

```{r lollipop_plot_8, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 8th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[8]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[8]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[8], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[9]) ) { names(mutRate.top.datasets)[9] }`

```{r lollipop_plot_9, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 9th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[9]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[9]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[9], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[10]) ) { names(mutRate.top.datasets)[10] }`

```{r lollipop_plot_10, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 10th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[10]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[10]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[10], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[11]) ) { names(mutRate.top.datasets)[11] }`

```{r lollipop_plot_11, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 11th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[11]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[11]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[11], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[12]) ) { names(mutRate.top.datasets)[12] }`

```{r lollipop_plot_12, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 12th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[12]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[12]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[12], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[13]) ) { names(mutRate.top.datasets)[13] }`

```{r lollipop_plot_13, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 13th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[13]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[13]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[13], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[14]) ) { names(mutRate.top.datasets)[14] }`

```{r lollipop_plot_14, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 14th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[14]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[14]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[14], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[15]) ) { names(mutRate.top.datasets)[15] }`

```{r lollipop_plot_15, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 15th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[15]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[15]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[15], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[16]) ) { names(mutRate.top.datasets)[16] }`

```{r lollipop_plot_16, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 16th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[16]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[16]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[16], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[17]) ) { names(mutRate.top.datasets)[17] }`

```{r lollipop_plot_17, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 17th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[17]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[17]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[17], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[18]) ) { names(mutRate.top.datasets)[18] }`

```{r lollipop_plot_18, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 18th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[18]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[18]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[18], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[19]) ) { names(mutRate.top.datasets)[19] }`

```{r lollipop_plot_19, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 19th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[19]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[19]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[19], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[20]) ) { names(mutRate.top.datasets)[20] }`

```{r lollipop_plot_20, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 20th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[20]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[20]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[20], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[21]) ) { names(mutRate.top.datasets)[21] }`

```{r lollipop_plot_21, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 21st top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[21]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[21]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[21], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[22]) ) { names(mutRate.top.datasets)[22] }`

```{r lollipop_plot_22, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 22nd top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[22]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[22]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[22], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[23]) ) { names(mutRate.top.datasets)[23] }`

```{r lollipop_plot_23, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 23rd top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[23]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[23]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[23], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[24]) ) { names(mutRate.top.datasets)[24] }`

```{r lollipop_plot_24, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 24th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[24]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[24]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[24], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[25]) ) { names(mutRate.top.datasets)[25] }`

```{r lollipop_plot_25, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 25th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[25]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[25]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[25], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[26]) ) { names(mutRate.top.datasets)[26] }`

```{r lollipop_plot_26, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 26th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[26]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[26]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[26], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[27]) ) { names(mutRate.top.datasets)[27] }`

```{r lollipop_plot_27, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 27th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[27]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[27]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[27], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[28]) ) { names(mutRate.top.datasets)[28] }`

```{r lollipop_plot_28, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 28th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[28]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[28]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[28], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[29]) ) { names(mutRate.top.datasets)[29] }`

```{r lollipop_plot_29, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 29th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[29]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[29]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[29], "is not available\n\n\n\n", sep=" "))
}
```

***

### `r if ( !is.na(names(mutRate.top.datasets)[30]) ) { names(mutRate.top.datasets)[30] }`

```{r lollipop_plot_30, comment = NA, message=FALSE, warning=FALSE, results = "asis"}
###### Generate lollipop plot for each dataset for 30th top mutated gene
if ( nrow(gff[HGNC %in% names(mutRate.top.datasets)[30]]) != 0 ) {
  try(lollipops.datasets(mafInfo = mafInfo, datasets  = datasets.list, gene  = names(mutRate.top.datasets)[30]), silent = TRUE)
} else {
   cat(paste("The protein structure for protein encoded by", names(mutRate.top.datasets)[30], "is not available\n\n\n\n", sep=" "))
}
```

***

## Mutation details {.tabset .tabset-fade}

Tables with detailed information, as as provided in corresponding MAF file(s), for the **`r top_genes_no[[i]]` most frequently mutated genes** (mutated in at least `r params$genes_min` % of patients) across all datasets, as well as for the **genes of interest** (if specified).

```{r details_mut_table, comment = NA, message=FALSE, warning=FALSE}
##### Provide detiles information for each dataset for user-defined number of top mutated genes
##### Create a new workbook
wb <- createWorkbook("MAF_top_mutations_details.xlsx")

##### Add worksheets, one for each dataset
for ( i in 1:length(mafFiles) ) {
  
  ##### Extract mutation details
  mut.details <- mafInfo[[i]]@data[ mafInfo[[i]]@data[, Hugo_Symbol] %in% names(mutRate.top.datasets), ]

  addWorksheet(wb, substring(datasets.list[i], 0, 31), tabColour = datasets.colour[[1]][i])
  writeData(wb, sheet = i, mut.details[ order(mut.details$Hugo_Symbol) , ])
}

saveWorkbook(wb, paste(outDir, "MAF_top_mutations_details.xlsx", sep="/"), overwrite = TRUE)
```

### Recurrently mutated genes

```{r details_mut, comment = NA, message=FALSE, warning=FALSE}
##### Provide detiles information for each dataset for user-defined number of top mutated genes
mut.details.datasets(mafInfo, datasets.list, names(mutRate.top.datasets)[c(1:max(unlist(top_genes_no[[i]])))])
##### Add extra lines to make sure that this section doesn't overlap with the next one
cat("\n\n\n")
```

***

### Additional genes of interest

```{r details_mut_goi, comment = NA, message=FALSE, warning=FALSE}
##### Provide detiles information for each dataset for user-defined number of top mutated genes

if ( !is.na(params$genes_list) ) {
  
  mut.details.datasets(mafInfo, datasets.list, unlist(genes_list_goi))
  
} else {
  cat("No list with genes of interest was provided\n\n\n\n")
}

##### Add extra lines to make sure that this section doesn't overlap with the next one
cat("\n\n\n")
```

***

## Heatmaps {.tabset .tabset-fade}

### Samples

Interactive heatmap(s) with colours indicating low (blue) and high (yellow) number of *various mutations types* (columns) detected in corresponding *samples* (rows). Samples are ordered by the number of mutations to facilitate identification of individuals with extreme mutation burden. 

```{r sample_summary_heatmap, message = FALSE, warning=FALSE, fig.width = 6, fig.height = 9}
suppressMessages(library(plotly))
suppressMessages(library(heatmaply))

##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

##### Display samples summary in a form of interactive heatmap
for ( i in 1:length(mafFiles) ) {

  sampleSummary <- data.frame(maftools::getSampleSummary(mafInfo[[i]]))
  rownames(sampleSummary) <-sampleSummary[,"Tumor_Sample_Barcode"]
  sampleSummary <- subset(sampleSummary, select=-c(Tumor_Sample_Barcode, total))

  ##### Generate interactive heatmap
  p <- heatmaply(sampleSummary, main = datasets.list[i], Rowv=NULL, Colv=NULL, scale="none", dendrogram="none", trace="none", hide_colorbar = FALSE, fontsize_row = 8, label_names=c("Sample","Mutation_type","Count")) %>%
  layout(width  = 900, height = 420, margin = list(l=150, r=10, b=80, t=50, pad=2), titlefont = list(size=16), xaxis = list(tickfont=list(size=10)), yaxis = list(tickfont=list(size=10)))

  ##### Add plot to the list for htmlwidgets
  widges.list[[i]] <- as_widget(ggplotly(p))

  ##### Save the heatmap as html (PLOTLY)
  htmlwidgets::saveWidget(as_widget(p), paste0(outDir, "/MAF_sample_summary_heatmap_", datasets.list[i], ".html"), selfcontained = TRUE)
}

##### Detach plotly package. Otherwise it clashes with other graphics devices
detach("package:heatmaply", unload=FALSE)
detach("package:plotly", unload=FALSE)

##### Print a list of htmlwidgets
widges.list
```

***

### Genes

Interactive heatmap(s) with colours indicating low (blue) and high (yellow) number of *various mutations types* (columns) detected in corresponding *genes* (rows). Genes are ordered by the number of reported mutations. The *total number of mutations* in individual genes, as well as the *number of samples with mutations* are also presented in the last three columns. Note, for transparency only the **`r top_genes_no[[i]]` most frequently mutated genes** (mutated in at least `r params$genes_min` % of patients) are presented.

```{r gene_summary_heatmap, message = FALSE, warning=FALSE, fig.width = 6, fig.height = 8}
suppressMessages(library(plotly))
suppressMessages(library(heatmaply))

##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

##### Display genes summary in a form of interactive heatmap
for ( i in 1:length(mafFiles) ) {

  geneSummary <- data.frame(maftools::getGeneSummary(mafInfo[[i]])[1:top_genes_no[[i]],])
  rownames(geneSummary) <-geneSummary[,"Hugo_Symbol"]
  geneSummary <- subset(geneSummary, select=-c(Hugo_Symbol))

  ##### Generate interactive heatmap
  p <- heatmaply(geneSummary, main = datasets.list[i], Rowv=NULL, Colv=NULL, scale="none", dendrogram="none", trace="none", hide_colorbar = FALSE, fontsize_row = 8, label_names=c("Gene","Mutation_type","Count")) %>%
  layout(width  = 900, height = 420, margin = list(l=150, r=10, b=80, t=50, pad=2), titlefont = list(size=16), xaxis = list(tickfont=list(size=10)), yaxis = list(tickfont=list(size=10)))
  
  ##### Add plot to the list for htmlwidgets
  widges.list[[i]] <- as_widget(ggplotly(p))

  ##### Save the heatmap as html (PLOTLY)
  htmlwidgets::saveWidget(as_widget(p), paste0(outDir, "/MAF_gene_summary_heatmap_", datasets.list[i], ".html"), selfcontained = TRUE)
}

##### Detach plotly package. Otherwise it clashes with other graphics devices
detach("package:heatmaply", unload=FALSE)
detach("package:plotly", unload=FALSE)

##### Print a list of htmlwidgets
widges.list
```

***

<!--This step is skipped for the moment as it is too computationally intensive.

### Pair-wise comparisons

Interactive heatmap of *adjusted p-values* from *Fisher Exact* test performed for all possible pair-wise comprisons to find differentially mutated genes between queried datasets and to aid identification of global differences in mutation patterns in corresponding MAF files. The rows and columns represent genes and datasets, respectively, and heatmap colours indicate the *Fisher Exact* test *adjusted p-values*. The colour key is on the left-hand side. **Low *p-values*** (yellow) **indicate differentially mutated genes between corresponding datasets**. Genes are clustered to facilite the indetification possible differences between individaul datasets. Note, only overlap of mutated genes reported across all MAF files is presented. This section is available for multiple MAFs and will be skipped if only one MAF file is provided.

--> 

```{r compare_mafs, message = FALSE, warning=FALSE, comment = NA, fig.width = 6, fig.height = 12, echo = FALSE, eval = FALSE}
##### Skip this section if only one MAF file is provided
if ( length(mafFiles) != 1 ) {
  
  suppressMessages(library(plotly))
  suppressMessages(library(heatmaply))
  
  #####  Create matrix of possible comparisons
  comb <- combn(levels(factor(datasets.list)), 2)
  
  #####  Get number of possible comparisons using the following formula:
  #
  # n!/((n-r)!(r!))
  #
  # n = the number of classes to compare
  # r = the number of elements for single comparison
  #
  ################################################################################
  
  combNo <- factorial(length(datasets.list))/(factorial(length(datasets.list)-2)*(factorial(2))) # n!/((n-r)!(r!))
  
  combNames <- NULL
  mafCompare.res <- vector("list", combNo)
  
  ##### Perform pair-wise datasets MAFs comparisons
  for (i in 1:combNo) {
  
    combNames[i] <- paste(comb[1,i], comb[2,i], sep=" vs ")
  
    mafCompare.res[[i]] <- maftools::mafCompare(m1 = mafInfo[[comb[1,i]]], m2 = mafInfo[[comb[2,i]]], m1Name = comb[1,i], m2Name = comb[2,i], minMut = 0)$results
  
    ##### Sort the data-frame by gene symbol
    mafCompare.res[[i]] <- mafCompare.res[[i]][order(mafCompare.res[[i]]$Hugo_Symbol)]
  
    rownames(mafCompare.res[[i]]) <- mafCompare.res[[i]]$Hugo_Symbol
  }
  
  ##### Extract the intersection of genes for the heatmap
  common_genes <-  Reduce(intersect, lapply(mafCompare.res, row.names))
  mafCompare.res.common <-  lapply(mafCompare.res, function(x) { x[row.names(x) %in% common_genes,] })
  
  ##### Extract adjusted p-values for each comparison, use genes names fow row names
  mafCompare.res.common.p <- do.call(cbind, lapply(mafCompare.res.common, function(x) x[, c("Hugo_Symbol", "adjPval")]))
  common_genes <- mafCompare.res.common.p$Hugo_Symbol
  mafCompare.res.common.p <- mafCompare.res.common.p[,-c("Hugo_Symbol")]
  mafCompare.res.common.p <- as.matrix(mafCompare.res.common.p)
  colnames(mafCompare.res.common.p) <- combNames
  rownames(mafCompare.res.common.p) <- common_genes
  
  ##### Cluster genes
  hr <- hclust(as.dist(dist(mafCompare.res.common.p, method="euclidean")), method="ward.D")
  
  ##### Generate interactive heatmap
  p <- heatmaply(mafCompare.res.common.p, Rowv=as.dendrogram(hr), Colv=NULL, viridis(n=256, alpha = 1, begin = 1, end = 0, option = "viridis"), scale="none", dendrogram="row", trace="none", hide_colorbar = FALSE, fontsize_row = 8, label_names=c("Gene","Comparison","P-value")) %>%
    layout(width  = 900, height = 900, margin = list(l=150, r=10, b=350, t=50, pad=2), titlefont = list(size=16), xaxis = list(tickfont=list(size=10)), yaxis = list(tickfont=list(size=10)))
  
  as_widget(ggplotly(p))
  p
  ##### Save the heatmap as html (PLOTLY)
  htmlwidgets::saveWidget(as_widget(p), paste0(outDir, "/MAF_pair-wise_comparisons_heatmap.html"), selfcontained = TRUE)

  ##### Detach plotly package. Otherwise it clashes with other graphics devices
  detach("package:heatmaply", unload=FALSE)
  detach("package:plotly", unload=FALSE)
  
} else {
  
  cat(paste("This section was skipped since MAF for only one dataset (", datasets.list[i], ") was provided.\n\n", sep=" "))
}
```

```{r maf_visualisation_pdf, comment=NA, message=FALSE, warning=FALSE}
###### Save plots into PDF files
###### Generate separate file with plots for each dataset
for ( i in 1:length(mafFiles) ) {

  pdf( file = paste(outDir, "/MAF_summary_", datasets.list[i], ".pdf", sep = "") )

  ##### Plotting MAF summary
  par(mar=c(4,4,2,0.5), oma=c(1.5,2,2,1))
  plotmafSummary(maf = mafInfo[[i]], top = top_genes_no[[i]], rmOutlier = TRUE, addStat = 'median', dashboard = TRUE, titvRaw = FALSE)
  mtext("MAF summary", outer=TRUE,  cex=1, line=-0.5)

  ##### Drawing oncoplots for the top mutated genes in each dataset that has > 1 sample with non-synonymous variants
  if ( top_genes_no[[i]] > 1 ) {
    
    ##### Drawing oncoplots for the top mutated genes in each dataset
    plot.new()
    par(mar=c(4,4,2,0.5), oma=c(1.5,2,2,1))
    oncoplot(maf = mafInfo[[i]], top = top_genes_no[[i]], fontSize = 12, colbar_pathway = TRUE, removeNonMutated = FALSE)
  }
  
  ##### Drawing oncoplots for additional genes of interest (if specified)
  if ( !is.na(params$genes_list) && length(goi) > 1 ) {

    ##### Drawing oncoplots for the top mutated genes in each dataset that has > 1 sample with non-synonymous variants
    if ( length(genes_list_goi[[i]]) > 1 ) {
      
      plot.new()
      par(mar=c(4,4,2,0.5), oma=c(1.5,2,2,1))
      maftools::oncoplot(maf = mafInfo[[i]], genes = genes_list_goi[[i]], fontSize = 12, colbar_pathway = TRUE, removeNonMutated = FALSE)
    }
  }
  
  ##### Drawing distribution plots of the transitions and transversions
  titv.info = titv(maf = mafInfo[[i]], plot = FALSE, useSyn = TRUE)

  try(plotTiTv(res = titv.info), silent = TRUE)
  mtext("Transition and transversions distribution", outer=TRUE,  cex=1, line=-1.5)

  ##### Compare mutation load against TCGA cohorts
  tcgaCompare_mod(maf = mafInfo[[i]], cohortName = datasets.list[i], primarySite=TRUE)
  mtext("Mutation load in TCGA cohorts", outer=TRUE,  cex=1, line=-0.5)
}

##### Add extra lines to make sure that this section doesn't overlap with the next one
```

***

## Addendum

<details>
<summary>Parameters</summary>
<font size="2">

```{r params_info, comment = NA}
for ( i in 1:length(params) ) {

  cat(paste("Parameter: ", names(params)[i], "\nValue: ", paste(unlist(params[i]), collapse = ","), "\n\n", sep=""))
}
```

</font>
</details>

<details>
<summary>Session info</summary>
<font size="2">

```{r sessioninfo, comment = NA}
devtools::session_info()
```

</font>
</details>
