---
title: "MAF files summary"
author: "UMCCR"
date: "`r format(Sys.time(), '%d %B, %Y')`"
params:
  maf_dir: 
  maf_files: 
  datasets: 
  out_dir: 
output:
  html_document:
    keep_md: yes
    code_folding: hide
    toc: true
    toc_float: true
---

```{r define_functions}
##### Define functions

##### Assign colours to different datasets. These colours will be used to distinguish tabs in generated excel summary spreadsheets
getDatasetsColours <- function(datasets) {
  
  ##### Predefined selection of colours for datasets
  datasets.colours <- c("dodgerblue","firebrick","lightslategrey","darkseagreen","orange","darkcyan","bisque", "coral2", "cadetblue3","red","blue","green")
  
  f.datasets <- factor(datasets)
  vec.datasets <- datasets.colours[1:length(levels(f.datasets))]
  datasets.colour <- rep(0,length(f.datasets))
  for(i in 1:length(f.datasets))
    datasets.colour[i] <- vec.datasets[ f.datasets[i]==levels(f.datasets)]
  
  return( list(vec.datasets, datasets.colour) )
}
```

```{r load_libraries, warning=FALSE}
suppressMessages(library(knitr))
suppressMessages(library(maftools))
suppressMessages(library(openxlsx))
suppressMessages(library(DT))
```

## Datasets

This report summarises and visualises mutation data in [Mutation Annotation Format](https://software.broadinstitute.org/software/igv/MutationAnnotationFormat) (MAF) files for the following dataset(s):

```{r datasets, comment = NA}
##### Present patient cohorts to be summarised
mafFiles <- params$maf_files
datasets.list <- params$datasets

datasets.df <- as.data.frame(cbind(datasets.list, unlist(lapply(strsplit(mafFiles, split='/', fixed=TRUE), tail, n = 1L))))
names(datasets.df) <- c("Dataset", "MAF file")

DT::datatable( data = datasets.df, filter = "none", options = list(pageLength = length(mafFiles), sDom  = '<"top">lrt<"bottom">ip') )

cat("These MAF files are located in the following directory: ", params$maf_dir, "\n\n")
```

```{r load_data, comment = NA, message=FALSE, warning=FALSE}
##### Read MAF files and put associated info into a list
##### Create a list to store MAF info for individual datasets
mafInfo <- vector("list", length(mafFiles))
names(mafInfo) <- datasets.list

for ( i in 1:length(mafFiles) ) {
  mafInfo[[i]] = read.maf(maf = mafFiles[i], verbose = FALSE)
}

##### Create directory for output files
outDir <- paste(params$maf_dir, params$out_dir, sep = "/")

if ( !file.exists(params$out_dir) ){
  dir.create(outDir, recursive=TRUE)
}
```

## Summary tables {.tabset}

### Overall summary

Table(s) with basic information about each dataset, including *NCBI build*, *number fo samples* and *mutated genes*, *number of different mutation types*, as well as the *total number of mutations*, as reported in corresponding MAF file(s).

```{r overll_summary, warning=FALSE}
##### Write overall summary into a file
##### Assign different colour to individual datasets
datasets.colour <- getDatasetsColours(datasets.list)

##### Create a new workbook
wb <- createWorkbook("MAF_summary.xlsx")

##### Add worksheets, one for each dataset
for ( i in 1:length(mafFiles) ) {
  
    addWorksheet(wb, datasets.list[i], tabColour = datasets.colour[[1]][i])
    writeData(wb, sheet = i, mafInfo[[i]]@summary)
}

saveWorkbook(wb, paste(outDir, "MAF_summary.xlsx", sep="/"), overwrite = TRUE)

##### Present a MAF file summary table in the html report
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

for ( i in 1:length(mafFiles) ) {
  widges.list[[i]] <- DT::datatable( data = mafInfo[[i]]@summary, caption = htmltools::tags$caption(style = 'caption-side: top; text-align: left;', htmltools::strong(datasets.list[i])), filter = "none", options = list(pageLength = nrow(mafInfo[[i]]@summary), sDom  = '<"top">lrt<"bottom">ip') )
}

##### Print a list of htmlwidgets
widges.list

##### Add extra lines to make sure that this section doesn't overlap with the next one
cat("\n\n\n")
```

```{r maf_fields, warning=FALSE}
##### Additionally ceate an excel spreadsheet listing all fields (columns) in the individaul MAF files
##### Create a new workbook
wb <- createWorkbook("MAF_fields.xlsx")

##### Add worksheets, one for each dataset
for ( i in 1:length(mafFiles) ) {
  
    addWorksheet(wb, datasets.list[i], tabColour = datasets.colour[[1]][i])
    writeData(wb, sheet = i, maftools::getFields(mafInfo[[i]]))
}

saveWorkbook(wb, paste(outDir, "MAF_fields.xlsx", sep="/"), overwrite = TRUE)
```

### Samples summary

Table(s) summarising samples in individual datasets. Each table contains per-sample information (rows) about *number of different types of mutations* (columns), as well as the *total number of mutations* reported in corresponding MAF file.

```{r sample_summary, warning=FALSE}
##### Write samples summary into a file
##### Create a new workbook
wb <- createWorkbook("MAF_sample_summary.xlsx")

##### Add worksheets, one for each dataset
for ( i in 1:length(mafFiles) ) {
  
    addWorksheet(wb, datasets.list[i], tabColour = datasets.colour[[1]][i])
    writeData(wb, sheet = i, maftools::getSampleSummary(mafInfo[[i]]))
}

saveWorkbook(wb, paste(outDir, "MAF_sample_summary.xlsx", sep="/"), overwrite = TRUE)

##### Present a sample table in the html report
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

for ( i in 1:length(mafFiles) ) {
  widges.list[[i]] <- DT::datatable( data = maftools::getSampleSummary(mafInfo[[i]]), caption = htmltools::tags$caption(style = 'caption-side: top; text-align: left;', htmltools::strong(datasets.list[i])), filter = "top", options = list(pageLength = 5) )
}

##### Print a list of htmlwidgets
widges.list

##### Add extra lines to make sure that this section doesn't overlap with the next one
cat("\n\n\n")
```

### Genes summary

Table(s) summarising mutated genes in individual datasets. Each table contains per-gene information (rows) about *number of different types of mutations* (columns), as well as the *total number of mutations* reported in corresponding MAF file. The last two columns contain the *number of samples with mutations/alterations* in the corresponding gene.

```{r gene_summary, warning=FALSE}
##### Write gene summary into a file
##### Create a new workbook
wb <- createWorkbook("MAF_gene_summary.xlsx")

##### Add worksheets, one for each dataset
for ( i in 1:length(mafFiles) ) {
  
    addWorksheet(wb, datasets.list[i], tabColour = datasets.colour[[1]][i])
    writeData(wb, sheet = i, maftools::getGeneSummary(mafInfo[[i]]))
}

saveWorkbook(wb, paste(outDir, "MAF_gene_summary.xlsx", sep="/"), overwrite = TRUE)

##### Present a gene table in the html report
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

for ( i in 1:length(mafFiles) ) {
  widges.list[[i]] <- DT::datatable(data = maftools::getGeneSummary(mafInfo[[i]]), caption = htmltools::tags$caption(style = 'caption-side: top; text-align: left;', htmltools::strong(datasets.list[i])), filter = "top", options = list(pageLength = 5) )
}

##### Print a list of htmlwidgets
widges.list

##### Add extra lines to make sure that this section doesn't overlap with the next one
cat("\n\n\n")
```


## Summary plots {.tabset .tabset-fade}

### MAF summary

A per-MAF file summary including *frequency of various mutation/SNV types/classes* (top panel), the *number of variants in each sample* as a stacked bar-plot (bottom-left) and *variant types* as a box-plot (bottom-middle), as well as the *frequency of different mutation types* for the **top 10 mutated genes** (bottom-right). The horizontal dashed line in stacked bar-plot represents median number of variants across the dataset.

```{r maf_summary_plot, comment = NA}
###### Generate separate plot for each dataset
for ( i in 1:length(mafFiles) ) {

  cat(paste(datasets.list[i], "\n\n", sep=" "))
  
  ##### Plotting MAF summary
  par(mar=c(4,4,2,0.5), oma=c(1.5,2,2,1))
  maftools::plotmafSummary(maf = mafInfo[[i]], rmOutlier = TRUE, addStat = 'median', dashboard = TRUE, titvRaw = FALSE)
  mtext("MAF summary", outer=TRUE,  cex=1, line=-0.5)
}
```

### Oncoplot

Oncoplot(s) illustrating different types of mutations observed across samples for the **10 most frequently mutated genes**. The side- and top bar-plots present the frequency of mutations in these genes and in individual samples, respectively.

```{r maf_oncoplot, comment = NA}
###### Generate separate plot for each dataset
for ( i in 1:length(mafFiles) ) {

  cat(paste(datasets.list[i], "\n\n", sep=" "))

  ##### Drawing oncoplots for the top 10 genes in each dataset
  plot.new()
  par(mar=c(4,4,2,0.5), oma=c(1.5,2,2,1))
  maftools::oncoplot(maf = mafInfo[[i]], top = 10, fontSize = 12)
}
```

### Transitions/transversions

Plots presenting the transitions and transversions distribution in individual dataset(s). In each panel, the box-plots show the *overall distribution* of the six different conversions (*C>A*, *C>G*, *C>T*, *T>C*, *T>A* and *T>G*)(top-left), and the transitions and transversions *frequency* (top-right). The stacked bar-plot (bottom) displays the *fraction* of the six different conversions in each sample.

```{r maf_TiTv_plot, comment = NA}
###### Generate separate plot for each dataset
###### Create empty list for TiTv info from each dataset
titv.info <- list()

for ( i in 1:length(mafFiles) ) {

  cat(paste(datasets.list[i], "\n\n", sep=" "))

  ##### Drawing distribution plots of the transitions and transversions
  titv.info[[datasets.list[i]]] <- maftools::titv(maf = mafInfo[[i]], plot = FALSE, useSyn = TRUE)

  maftools::plotTiTv(res = titv.info[[datasets.list[i]]])
  mtext("Transition and transversions distribution", outer=TRUE,  cex=1, line=-1.5)
}
```

```{r maf_TiTv_table}
##### Write per-sample transitions and transversions distribution into a file
##### Create a new workbook
wb <- createWorkbook("MAF_summary_titv.xlsx")

##### Add worksheets, three for each dataset
for ( i in 1:length(mafFiles) ) {
  
    addWorksheet(wb, substring(paste0(datasets.list[i], " (fraction)"), 0, 31), tabColour = datasets.colour[[1]][i])
    writeData(wb, sheet = i*3-2, titv.info[[datasets.list[i]]]$fraction.contribution)
    
    addWorksheet(wb, substring(paste0(datasets.list[i], " (count)"), 0, 31), tabColour = datasets.colour[[1]][i])
    writeData(wb, sheet = i*3-1, titv.info[[datasets.list[i]]]$raw.counts)
    
    addWorksheet(wb, substring(paste0(datasets.list[i], " (TiTv fractions)"), 0, 31), tabColour = datasets.colour[[1]][i])
    writeData(wb, sheet = i*3, titv.info[[datasets.list[i]]]$TiTv.fractions)
}

saveWorkbook(wb, paste(outDir, "MAF_summary_titv.xlsx", sep="/"), overwrite = TRUE)
```

### Pair-wise comparisons

Interactive heatmap of *adjusted p-values* from *Fisher Exact* test performed for all possible pair-wise comprisons to find differentially mutated genes between queried datasets and to aid identification of global differences in mutation patterns in corresponding MAF files. The rows and columns represent genes and datasets, respectively, and heatmap colours indicate the *Fisher Exact* test *adjusted p-values*. The colour key is on the left-hand side. **Low *p-values*** (yellow) indicate differentially mutated genes between the corresponding datasets. Genes are clustered to facilite the indetification possible differences between individaul datasets. Note, only overlap of mutated genes reported across all MAF files is presented.

```{r compare_mafs, message = FALSE, warning=FALSE, fig.width = 6, fig.height = 12}
suppressMessages(library(plotly))
suppressMessages(library(heatmaply))

#####  Create matrix of possible comparisons
comb <- combn(levels(factor(datasets.list)), 2)

#####  Get number of possible comparisons using the following formula:
#
# n!/((n-r)!(r!))
#
# n = the number of classes to compare
# r = the number of elements for single comparison
#
################################################################################

combNo <- factorial(length(datasets.list))/(factorial(length(datasets.list)-2)*(factorial(2))) # n!/((n-r)!(r!))

combNames <- NULL
mafCompare.res <- vector("list", combNo)

##### Perform pair-wise datasets MAFs comparisons
for (i in 1:combNo) {

  combNames[i] <- paste(comb[1,i], comb[2,i], sep=" vs ")

  mafCompare.res[[i]] <- maftools::mafCompare(m1 = mafInfo[[comb[1,i]]], m2 = mafInfo[[comb[2,i]]], m1Name = comb[1,i], m2Name = comb[2,i], minMut = 0)$results

  ##### Sort the data-frame by gene symbol
  mafCompare.res[[i]] <- mafCompare.res[[i]][order(mafCompare.res[[i]]$Hugo_Symbol)]

  rownames(mafCompare.res[[i]]) <- mafCompare.res[[i]]$Hugo_Symbol
}

##### Extract the intersection of genes for the heatmap
common_genes <-  Reduce(intersect, lapply(mafCompare.res, row.names))
mafCompare.res.common <-  lapply(mafCompare.res, function(x) { x[row.names(x) %in% common_genes,] })

##### Extract adjusted p-values for each comparison, use genes names fow row names
mafCompare.res.common.p <- do.call(cbind, lapply(mafCompare.res.common, function(x) x[, c("Hugo_Symbol", "adjPval")]))
common_genes <- mafCompare.res.common.p$Hugo_Symbol
mafCompare.res.common.p <- mafCompare.res.common.p[,-c("Hugo_Symbol")]
mafCompare.res.common.p <- as.matrix(mafCompare.res.common.p)
colnames(mafCompare.res.common.p) <- combNames
rownames(mafCompare.res.common.p) <- common_genes

##### Cluster genes
hr <- hclust(as.dist(dist(mafCompare.res.common.p, method="euclidean")), method="ward.D")

##### Generate interactive heatmap
p <- heatmaply(mafCompare.res.common.p, Rowv=as.dendrogram(hr), Colv=NULL, viridis(n=256, alpha = 1, begin = 1, end = 0, option = "viridis"), scale="none", dendrogram="row", trace="none", hide_colorbar = FALSE, fontsize_row = 8, label_names=c("Gene","Comparison","P-value")) %>%
  layout(width  = 900, height = 900, margin = list(l=150, r=10, b=250, t=50, pad=2), titlefont = list(size=16), xaxis = list(tickfont=list(size=10)), yaxis = list(tickfont=list(size=10)))

as_widget(ggplotly(p))

##### Save the heatmap as html (PLOTLY)
htmlwidgets::saveWidget(as_widget(p), paste0(outDir, "/MAF_pair-wise_comparisons_heatmap.html"), selfcontained = TRUE)

##### Detach plotly package. Otherwise it clashes with other graphics devices
detach("package:heatmaply", unload=FALSE)
detach("package:plotly", unload=FALSE)
```

### Cross-cancers comparison

Plot(s) illustrating the mutation load in investigated dataset(s) along with distribution of variants compiled from over 10,000 whole-exome sequencing samples across 33 [TCGA](https://cancergenome.nih.gov/) landmark cohorts. Every dot represents a sample whereas the red horizontal lines are the median numbers of mutations in the respective cancer types. The vertical axis (log scaled) shows the number of mutations per megabase whereas the different cancer types are ordered on the horizontal axis based on their median numbers of somatic mutations. This plot is similar to the one described in the paper [Signatures of mutational processes in human cancer](https://www.ncbi.nlm.nih.gov/pubmed/23945592) by Alexandrov *et al*.

```{r maf_tcga_cohorts, comment = NA, message=FALSE, warning=FALSE}
###### Generate separate plot for each dataset
for ( i in 1:length(mafFiles) ) {
  
  cat(paste(datasets.list[i], "\n\n", sep=" "))
  
  ##### Compare mutation load against TCGA cohorts
  maftools::tcgaCompare(maf = mafInfo[[i]], cohortName = datasets.list[i], primarySite=TRUE)
}
```

## Heatmaps {.tabset .tabset-fade}

### Samples

Interactive heatmap(s) with colours indicating low (blue) and high (yellow) number of *various mutations types* (columns) detected in corresponding *samples* (rows). Samples are ordered by the number of mutations to facilitate identification of individuals with extreme mutation burden. 

```{r sample_summary_heatmap, message = FALSE, warning=FALSE, fig.width = 6, fig.height = 9}
suppressMessages(library(plotly))
suppressMessages(library(heatmaply))

##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

##### Display samples summary in a form of interactive heatmap
for ( i in 1:length(mafFiles) ) {

  sampleSummary <- data.frame(maftools::getSampleSummary(mafInfo[[i]]))
  rownames(sampleSummary) <-sampleSummary[,"Tumor_Sample_Barcode"]
  sampleSummary <- subset(sampleSummary, select=-c(Tumor_Sample_Barcode, total))

  ##### Generate interactive heatmap
  p <- heatmaply(sampleSummary, main = datasets.list[i], Rowv=NULL, Colv=NULL, scale="none", dendrogram="none", trace="none", hide_colorbar = FALSE, fontsize_row = 8, label_names=c("Sample","Mutation_type","Count")) %>%
  layout(width  = 900, height = 550, margin = list(l=150, r=10, b=230, t=50, pad=2), titlefont = list(size=16), xaxis = list(tickfont=list(size=10)), yaxis = list(tickfont=list(size=10)))

  ##### Add plot to the list for htmlwidgets
  widges.list[[i]] <- as_widget(ggplotly(p))

   ##### Save the heatmap as html (PLOTLY)
  htmlwidgets::saveWidget(as_widget(p), paste0(outDir, "/MAF_sample_summary_heatmap_", datasets.list[i], ".html"), selfcontained = TRUE)

  ##### Plotly option
  #p <- plot_ly(x = colnames(sampleSummary), y = rownames(sampleSummary), z = as.matrix(sampleSummary), height = 600, type = "heatmap") %>%
  #layout(title = paste0("Samples summary: ", datasets.list[i]), autosize = TRUE, margin = list(l=150, r=10, b=100, t=100, pad=4), showlegend = TRUE)

  #widges.list[[i]] <- ggplotly(p)
}

##### Detach plotly package. Otherwise it clashes with other graphics devices
detach("package:heatmaply", unload=FALSE)
detach("package:plotly", unload=FALSE)

##### Print a list of htmlwidgets
widges.list
```

### Genes

Interactive heatmap(s) with colours indicating low (blue) and high (yellow) number of *various mutations types* (columns) detected in corresponding *genes* (rows). Genes are ordered by the number of reported mutations. The *total number of mutations* in individual genes, as well as the *number of samples with mutations* are also presented in the last three columns. Note, for transparency only the **top 50 mutated genes** are presented.

```{r gene_summary_heatmap, message = FALSE, warning=FALSE, fig.width = 6, fig.height = 8}
suppressMessages(library(plotly))
suppressMessages(library(heatmaply))

##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

##### Display genes summary in a form of interactive heatmap
for ( i in 1:length(mafFiles) ) {

  geneSummary <- data.frame(maftools::getGeneSummary(mafInfo[[i]])[1:50,])
  rownames(geneSummary) <-geneSummary[,"Hugo_Symbol"]
  geneSummary <- subset(geneSummary, select=-c(Hugo_Symbol))

  ##### Generate interactive heatmap
  p <- heatmaply(geneSummary, main = datasets.list[i], Rowv=NULL, Colv=NULL, scale="none", dendrogram="none", trace="none", hide_colorbar = FALSE, fontsize_row = 8, label_names=c("Gene","Mutation_type","Count")) %>%
  layout(width  = 900, height = 550, margin = list(l=150, r=10, b=230, t=50, pad=2), titlefont = list(size=16), xaxis = list(tickfont=list(size=10)), yaxis = list(tickfont=list(size=10)))

  ##### Add plot to the list for htmlwidgets
  widges.list[[i]] <- as_widget(ggplotly(p))

  ##### Save the heatmap as html (PLOTLY)
  htmlwidgets::saveWidget(as_widget(p), paste0(outDir, "/MAF_gene_summary_heatmap_", datasets.list[i], ".html"), selfcontained = TRUE)

  ##### Plotly option
  #p <- plot_ly(x = colnames(geneSummary), y = rownames(geneSummary), z = as.matrix(geneSummary), height = 600, type = "heatmap") %>%
  #layout(title = paste0("Genes summary: ", datasets.list[i]), autosize = TRUE, margin = list(l=150, r=10, b=100, t=100, pad=4), showlegend = TRUE)

  #widges.list[[i]] <- ggplotly(p)
}

##### Detach plotly package. Otherwise it clashes with other graphics devices
detach("package:heatmaply", unload=FALSE)
detach("package:plotly", unload=FALSE)

##### Print a list of htmlwidgets
widges.list
```

```{r maf_visualisation_pdf, comment = NA, message=FALSE, warning=FALSE}
###### Save plots into PDF files
###### Generate separate file with plots for each dataset
for ( i in 1:length(mafFiles) ) {

  pdf( file = paste(outDir, "/MAF_summary_", datasets.list[i], ".pdf", sep = "") )

  ##### Plotting MAF summary
  par(mar=c(4,4,2,0.5), oma=c(1.5,2,2,1))
  plotmafSummary(maf = mafInfo[[i]], rmOutlier = TRUE, addStat = 'median', dashboard = TRUE, titvRaw = FALSE)
  mtext("MAF summary", outer=TRUE,  cex=1, line=-0.5)

  ##### Drawing oncoplots for the top 10 genes in each dataset
  plot.new()
  par(mar=c(4,4,2,0.5), oma=c(1.5,2,2,1))
  oncoplot(maf = mafInfo[[i]], top = 10, fontSize = 12)

  ##### Drawing distribution plots of the transitions and transversions
  titv.info = titv(maf = mafInfo[[i]], plot = FALSE, useSyn = TRUE)

  plotTiTv(res = titv.info)
  mtext("Transition and transversions distribution", outer=TRUE,  cex=1, line=-1.5)

  ##### Compare mutation load against TCGA cohorts
  #maftools::tcgaCompare(maf = mafInfo[[i]], cohortName = datasets.list[i], primarySite=TRUE)
  #mtext("Mutation load in TCGA cohorts", outer=TRUE,  cex=1, line=-0.5)

  dev.off()
}

##### Add extra lines to make sure that this section doesn't overlap with the next one
```

## Parameters

```{r params_info, comment = NA}
for ( i in 1:length(params) ) {

  cat(paste("Parameter: ", names(params)[i], "\nValue: ", paste(unlist(params[i]), collapse = ","), "\n\n", sep=""))
}
```

## Addendum

```{r sessioninfo}
devtools::session_info()
```
